filesource/filesource.go

```diff
@@ -402,42 +422,17 @@ func (r *reader) makeParseConfig(obj ObjectInfo) *parser.Config {
 }
 
 func (r *reader) emit(lines []json.RawMessage) error {
-	// Message which wraps Records we'll generate.
-	var wrapper = &airbyte.Message{
-		Type: airbyte.MessageTypeRecord,
-		Record: &airbyte.Record{
-			Stream:    r.prefix,
-			EmittedAt: time.Now().UTC().UnixNano() / int64(time.Millisecond),
-		},
-	}
-	// Message which wraps State checkpoints we'll generate.
-	// Use a custom struct that matches the shape of airbyte.Message,
-	// but specializes State.Data to our type and avoids a double-encoding.
-	var stateWrapper = &struct {
-		Type  airbyte.MessageType `json:"type"`
-		State struct {
-			Data States `json:"data"`
-		} `json:"state,omitempty"`
-	}{
-		Type: airbyte.MessageTypeState,
-	}
-
-	r.shared.mu.Lock()
-	defer r.shared.mu.Unlock()
-
-	// Write out all records, wrapped in |wrapper|.
 	for _, line := range lines {
-		wrapper.Record.Data = line
-		if err := r.shared.enc.Encode(wrapper); err != nil {
+		if err := r.stream.Documents(r.binding, line); err != nil {
 			return err
 		}
 	}
 
-	// Update and write out state.
 	r.shared.states[r.prefix] = r.state
```

Comment:
I think we still need a lock here for concurrent access of multiple bindings to `r.shared.states`. And along with that, the shared mutex used by all bindings that was previously made by `r.shared.mu = sharedMu` in `Source.Read`.

As an alternative - could an individual binding emit its state checkpoint as a merge patch? As I understand it the state is a [map of "prefixes" to their state](https://github.com/estuary/connectors/blob/cc330e5d80cbdf938c49d9887a623949cbe43668/filesource/state.go#L12). So an individual binding ought to be able to track its own state without any concurrency issues and checkpoint with an appropriate merge patch. I'll caveat by saying I'm not 100% sure if this would work, since I don't fully understand why the merge checkpoint wasn't working before even with the full state object and it works without the merge - but I can see that it does.

Reply:
@williamhbaker Okay I dug a bit deeper into this and used https://github.com/estuary/flow/pull/1060 to troubleshoot this.

Turns out, we emit full state objects, until we are finished with the sweep, then we only emit a state object with `minBound` as its only property. With a merge patch, the other fields stay on, which is not what we want, we want to clear all other properties (that's how the connector expects to receive the checkpoint).

One way to do this, is to have an explicit `null` for all the missing attributes, and that can be achieved by removing `omitempty` json tag from attributes of `State`. This way, we can still emit merge patch to set checkpoint of each prefix independently, but we will also make sure that if our `State` has `nil` values, we actually explicitly send a `null` json value, which in JSON Merge Patch RFC, merging a property with an explicit `null` value just removes that property, exactly what we want.

I tested this, seems to work as expected now, see my last commit 👍🏽 

```
Merge Patch for Checkpoint: {
  "daves-test-bucket-1234562/": {
    "complete": true,
    "maxBound": "2023-05-26T13:25:49Z",
    "maxMod": "2023-05-26T13:23:54Z",
    "minBound": "0001-01-01T00:00:00Z",
    "path": "daves-test-bucket-1234562/Test Spreadsheet - Sheet1.csv"
  }
}
Checkpoint: {
  "daves-test-bucket-1234562/": {
    "complete": true,
    "maxBound": "2023-05-26T13:25:49Z",
    "maxMod": "2023-05-26T13:23:54Z",
    "minBound": "0001-01-01T00:00:00Z",
    "path": "daves-test-bucket-1234562/Test Spreadsheet - Sheet1.csv",
    "records": 5
  }
}
{"ts":"2023-05-26T13:26:20.807749959+00:00","level":"warn","message":"\u001b[36mINFO\u001b[0m[0002] completed sweep of daves-test-bucket-1234562/ from 0001-01-01 00:00:00 +0000 UTC through 2023-05-26 13:25:49 +0000 UTC"}
Merge Patch for Checkpoint: {
  "daves-test-bucket-1234562/": {
    "minBound": "2023-05-26T13:23:54Z"
  }
}
Checkpoint: {
  "daves-test-bucket-1234562/": {
    "complete": true,
    "maxBound": "2023-05-26T13:25:49Z",
    "maxMod": "2023-05-26T13:23:54Z",
    "minBound": "2023-05-26T13:23:54Z",
    "path": "daves-test-bucket-1234562/Test Spreadsheet - Sheet1.csv",
    "records": 5
  }
}
```